app/domain/authentication/authn_k8s/execute_command_in_container.rb

```diff
@@ -25,43 +28,30 @@ module AuthnK8s
       DEFAULT_KUBE_EXEC_COMMAND_TIMEOUT = 5
 
       def call
-        @message_log    = MessageLog.new
-        @channel_closed = false
-
-        url       = server_url(@cmds, @stdin)
-        headers   = kube_client.headers.clone
-        ws_client = WebSocket::Client::Simple.connect(url, headers: headers)
-
-        add_websocket_event_handlers(ws_client, @body, @stdin)
-
+        init_ws_client
+        add_websocket_event_handlers(ws_client)
         wait_for_close_message
-
-        unless @channel_closed
-          raise Errors::Authentication::AuthnK8s::CommandTimedOut.new(
-            timeout,
-            @container,
-            @pod_name
-          )
-        end
-
-        # TODO: raise an `WebsocketServerFailure` here in the case of ws :error
-
-        @message_log.messages
+        verify_channel_is_closed
+        verify_error_stream_is_empty
+        websocket_messages
       end
 
       def on_open(ws_client, body, stdin)
         hs       = ws_client.handshake
         hs_error = hs.error
 
         if hs_error
-          ws_client.emit(:error, "Websocket handshake error: #{hs_error.inspect}")
+          ws_client.emit(
+            :error,
+            Errors::Authentication::AuthnK8s::WebSocketHandshakeError.new(hs_error.inspect)
+          )
```

Comment:
Is this correct?  It looks like we're sending the error back through the websocket connection rather than raising it ourselves.  Is this intentional?  If so, can you explain what's happening?

Also I was confused because `emit` didn't seem to be documented for this websocket client:

https://rdoc.info/github/imanel/websocket-eventmachine-client/master/frames#simple-client-example

Reply:
i added this comment:
```
          # Add the error to the stream instead of raising it so we continue to
          # receive other messages. At the end of this class run we will verify
          # the stream to see if any errors were written
```

Reply:
removed this, please see in commits